src/Storages/StorageDuckDB.cpp

```diff
@@ -0,0 +1,183 @@
+#include <Storages/transformQueryForExternalDatabase.h>
+#include <Storages/checkAndGetLiteralArgument.h>
+#include <QueryPipeline/Pipe.h>
+#include <Common/filesystemHelpers.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int DUCKDB_ENGINE_ERROR;
+}
+
+StorageDuckDB::StorageDuckDB(
+    const StorageID & table_id_,
+    DuckDBPtr duckdb_instance_,
+    const String & database_path_,
+    const String & remote_table_name_,
+    const ColumnsDescription & columns_,
+    const ConstraintsDescription & constraints_,
+    ContextPtr context_)
+    : IStorage(table_id_)
+    , WithContext(context_->getGlobalContext())
+    , remote_table_name(remote_table_name_)
+    , database_path(database_path_)
+    , duckdb_instance(duckdb_instance_)
+    , log(&Poco::Logger::get("StorageDuckDB (" + table_id_.table_name + ")"))
+{
+    StorageInMemoryMetadata storage_metadata;
+    storage_metadata.setColumns(columns_);
+    storage_metadata.setConstraints(constraints_);
+    setInMemoryMetadata(storage_metadata);
+}
+
+
+Pipe StorageDuckDB::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context_,
+    QueryProcessingStage::Enum,
+    size_t max_block_size,
+    size_t /*num_streams*/)
+{
+    if (!duckdb_instance)
+        duckdb_instance = openDuckDB(database_path, getContext(), /* throw_on_error */true);
+
+    storage_snapshot->check(column_names);
+
+    String query = transformQueryForExternalDatabase(
+        query_info,
+        storage_snapshot->metadata->getColumns().getOrdinary(),
+        IdentifierQuotingStyle::DoubleQuotes,
+        "",
+        remote_table_name,
+        context_);
+    LOG_TRACE(log, "Query: {}", query);
+
+    Block sample_block;
+    for (const String & column_name : column_names)
+    {
+        auto column_data = storage_snapshot->metadata->getColumns().getPhysical(column_name);
+        sample_block.insert({column_data.type, column_data.name});
+    }
+
+    return Pipe(std::make_shared<DuckDBSource>(duckdb_instance, query, sample_block, max_block_size));
+}
+
+
+class DuckDBSink : public SinkToStorage
+{
+public:
+    explicit DuckDBSink(
+        const StorageDuckDB & storage_,
+        const StorageMetadataPtr & metadata_snapshot_,
+        StorageDuckDB::DuckDBPtr duckdb_instance_,
+        const String & remote_table_name_)
+        : SinkToStorage(metadata_snapshot_->getSampleBlock())
+        , storage{storage_}
+        , metadata_snapshot(metadata_snapshot_)
+        , duckdb_instance(duckdb_instance_)
+        , remote_table_name(remote_table_name_)
+    {
+    }
+
+    String getName() const override { return "DuckDBSink"; }
+
+    void consume(Chunk chunk) override
+    {
+        auto block = getHeader().cloneWithColumns(chunk.getColumns());
+        WriteBufferFromOwnString buf;
+
+        buf << "INSERT INTO ";
```

Comment:
The're API methods to perform more efficient data loading https://duckdb.org/docs/data/overview#appender-c-and-java 

Maybe it's ok to use INSERT for the first version (in that case we can just leave todo comment), but according to docs `Appender` API is preferable